### DISCUSSION TITLE
Pickle server

### DISCUSSION THREAD

**Discussion by roin93** [FreeOpcUa/opcua-asyncio#511] [Q&A]
Hello all together,

my current project constists of creating many, many nodes to a server. The normal startup of my server takes around 10 to 15 minutes and I want to use the same code for another 20 instances. So the final time I need to startup my server setup would take several hours.

Does one of you tried to use pickle for the server?
I would create all the nodes I need for the whole server stuff and save / load a pickle.
I know of the possibility to export and import xml files but it would take again lots of processing time, right?

I currently use the sync wrapper for my project and would like to just load a saved pickle of the server. I did not try it yet, just wanted to ask if there is some experience about that.

**Comment by swamper123** [comment]
Have a look here:
https://github.com/FreeOpcUa/opcua-asyncio/blob/12c0e7a05d756219ac56448edf7f78a030a3d049/asyncua/server/address_space.py#L589

There are already mechanism for such a purpose (but I don't know how good they work atm). Perhaps this is something you are looking for.

Just out of interest: How many nodes are you talking about? Round about?

**Reply by roin93** [reply]
At the moment I create many VariableTypes from Siemens at my server. This alone are more than 5000 (sub-) nodes.
Then I create all existing "real" information nodes which are roughly about 2000 Nodes. 
But that is only one machine... I want to do it for like 20 more machines, if possible within this script...

Im currently developing the script and have to restart it many times. Waiting for such a long time is just not practical anymore.

**Reply by roin93** [reply]
This seems not to work.
I tried to call 
server.address_space.AddressSpace.dump() but I just get the error that server does not have no attribute like address_space.
I am not familiar with the Sync-Wrapper and im not sure how I can just add the address_space attribute to it to get it from its async class.

Also I tried to just pickle the server object itself but I get the error that I cannot pickle thread._lock.
Is there an example on startup performance when I normally would build my node tree from several loops?

**Reply by swamper123** [reply]
Have you tried these from iserver class?

https://github.com/FreeOpcUa/opcua-asyncio/blob/0a1cce0783d93493515ff92e55952e63909fc834/asyncua/server/internal_server.py#L150

https://github.com/FreeOpcUa/opcua-asyncio/blob/0a1cce0783d93493515ff92e55952e63909fc834/asyncua/server/internal_server.py#L156

**Reply by roin93** [reply]
In fact I did not use the iserver-class. Good point!
